### ISSUE TITLE
Collapse folder in file explorer

### ISSUE THREAD

**Issue by balaji-sivasakthi** [zed-industries/zed#6880]
### Check for existing issues

- [X] Completed

### Describe the feature

When I've big file tree, I need to close the all the opened child directories, this feature really helps when it comes to handy.

### If applicable, add mockups / screenshots to help present your vision of the feature

<img width="54" alt="Screenshot 2024-01-27 at 9 51 02 PM" src="https://github.com/zed-industries/zed/assets/65065614/6a3a2877-6688-4ec2-b0ab-cca3338ad8ce">

**Comment by d1y** [comment]
Is this what you're talking about?
```json
{
  "context": "ProjectPanel && not_editing",
  "bindings": {
    "m": "project_panel::CollapseAllEntries"
  }
}
```
![hhh](https://github.com/zed-industries/zed/assets/45585937/19fc668b-1a8d-4809-9b43-2a1597a10496)

**Comment by balaji-sivasakthi** [comment]
Yes, @d1y

**Comment by itsmewes** [comment]
This is great, would be nice to have an option to only close the children of the folder that is currently active. So if we have the below file structure where all folders are open and `Folder 1` is selected in the side panel, running a collapse command on it would only collapse `Folder 1` and `Folder 1.2` but would leave `Folder 2` and it's children open.

Project
 - Folder 1
   - Folder 1.1
     - file1.txt
 - Folder 2
   - Folder 2.1
     - file2.txt

`project_panel::CollapseAllEntries` fixes 99% of my use cases but it is nice to have the above functionality as well if possible one day 😊

**Comment by TheGlorySaint** [comment]
I'd love to see not only having a Key bind to close the Folders, but also to have a Icon to do it.

**Comment by orbiteleven** [comment]
Yeah, it's like there needs to be an option like `project_panel::CollapseAllEntries` but only for the selected folder and down, so like `project_panel::CollapseAllEntriesBelowSelection` or something.

**Comment by gh0stonio** [comment]
> Yeah, it's like there needs to be an option like `project_panel::CollapseAllEntries` but only for the selected folder and down, so like `project_panel::CollapseAllEntriesBelowSelection` or something.

Agree with this. Having an option to be able to collapse a section using a `Cmd+click` like on VsCode could be really useful.

**Comment by edtorba** [comment]
> I'd love to see not only having a Key bind to close the Folders, but also to have a Icon to do it.

+1

**Comment by segersniels** [comment]
Yeah definitely would love to have a button to collapse my entire directory tree. Especially helpful in big monorepos where you need to navigate between workspaces frequently.

**Comment by tomatau** [comment]
Is it possible to trigger this action from inside the editor pane? Right now I need to be focused in the LeftDock for this command to work

**Comment by AdamSchinzel** [comment]
This would be very nice, I use it on a daily basis

**Comment by AdamSchinzel** [comment]
That would be very nice, I use it every day.

**Comment by cabaucom376** [comment]
Im not sure if there's a better issue to comment on, but support for `option + clicking` folders in the explorer to toggle nested folders expansion state would be an amazing quality of life improvement.


https://github.com/zed-industries/zed/assets/104145051/936d0a2b-e1fc-4687-b56d-4389ce4e9400

**Comment by JosephTLyons** [comment]
Are we able to close this issue @balaji-sivasakthi? It seems like @d1y gave you a snippet that works for you? Or is there something else we need to do here? As for the follow-up requests - let's open new issues for those so that we can track interest on those items individually.

**Comment by edtorba** [comment]
> Are we able to close this issue @balaji-sivasakthi? It seems like @d1y gave you a snippet that works for you? Or is there something else we need to do here? As for the follow-up requests - let's open new issues for those so that we can track interest on those items individually.

Adding the actual icon as shown in the screenshot would be nice. To be honest, many people won't find this issue and won't know about the key binding, imo.

**Comment by balaji-sivasakthi** [comment]
@JosephTLyons Yeah, key binding will work, but how will a new user realize that? Whether they search for this issue again or not, I think some people raise this issue often. So, my point of view is to bring that icon back, which will be more appropriate.

**Comment by JosephTLyons** [comment]
> @JosephTLyons Yeah, key binding will work, but how will a new user realize that? Whether they search for this issue again or not, I think some people raise this issue often. So, my point of view is to bring that icon back, which will be more appropriate.

Point noted - I rebranded this issue to be a discoverability issue and will leave it open.

**Comment by sant123** [comment]
I have created this keymap but only works when focusing the project panel.

```json
{
    "context": "ProjectPanel",
    "bindings": {
      "ctrl-alt-m": "project_panel::CollapseAllEntries"
    }
  }
```

Can this work anywhere?

**Comment by jerryjappinen** [comment]
Some common patterns from other apps (I've synced this behavior across design apps and code editors):

- Alt/Option click folder in file tree: close/open all descendants of that folder (no need for keyboard shortcut or icon for a given selection)
- Alt/Option + L: close/collapse all open layers/folders in file tree (except the root level, if one exists). This one is handy to just generally "reset" the stage while working and browsing

The former is, in my opinion, kindof no-brainer addition that doesn't come in the way anywhere else, and common behavior across apps. The latter, I guess, is already possible in key bindings - but I wasn't able to get it working.

**Comment by onx2** [comment]
Closing folders within the folder you close is the default behavior IIRC in JetBrains Rider, I really like that behavior and +1 the option to have it in Zed since I'm trying to make it my primary IDE. 😄

**Comment by brazillierjo** [comment]
Right click on the top folder structure -> "Collapse all"

**Comment by carhartl** [comment]
@sant123 You can accomplish this with [key remapping](https://zed.dev/docs/key-bindings#remapping-keys):

```json
  {
    "context": "Editor",
    "bindings": {
      "cmd-alt-r": ["workspace::SendKeystrokes", "cmd-shift-e c"]
    }
  }
```

(`cmd-shift-e` is, I believe, the VS Code keymap for toggling the project panel focus, so depending on what you have set up might be a different one...)

**Comment by devnshankar** [comment]
This feature really needs to be implemented working on bigger projects with huge files and folder structure becomes messy sometimes

**Comment by Gptimus** [comment]
@d1y can you share your theme, please ?

**Comment by Koshkaj** [comment]
I have implemented this at #60280, although not really sure if this'll be considered to be merged ¯\_(ツ)_/¯

**Comment by danilo-leal** [comment]
Heya everyone, chiming in here given we've had a few PRs over the years attempting to add buttons to the project panel for this and other actions. I gave @Koshkaj a reply that I think sort of serves as rationale for why I think we're good to close this issue: https://github.com/zed-industries/zed/pull/60280#issuecomment-4866495931. This issue was open more than two years ago, and ever since, this and many other actions have been included in the project panel's right-click context menu with their default keybindings displayed. To preserve Zed's aesthetic and minimalism, I think this should be enough for now. I'd risk betting that trying a right-click in the project panel to find these actions is an intuitive thing; something most developers would naturally do, and therefore, I think we'll be alright without a button just for this. Open to feedback if anyone feels strongly in opposition. Thanks!
